pkg/server/registry/images/strategy.go

```diff
@@ -3,177 +3,156 @@ package images
+		if !(includesTag(tag) || includesRepoAndTag(tag)) {
+			duplicateTag[tag+":latest"] = true
+		} else {
+			duplicateTag[tag] = true
 		}
-		result.Items = append(result.Items, list.Items...)
-	}
-
-	return result, nil
-}
 
-func (s *Strategy) forNamespace(ctx context.Context, ns string) (apiv1.ImageList, error) {
-	if ok, err := buildkit.Exists(ctx, s.client); err != nil {
-		return apiv1.ImageList{}, err
-	} else if !ok {
-		return apiv1.ImageList{}, nil
-	}
-
-	opts, err := remoteopts.WithServerDialer(ctx, s.client)
-	if err != nil {
-		return apiv1.ImageList{}, err
 	}
+	imageList := &apiv1.ImageList{}
 
-	repo, err := getRepo(ns)
-	if err != nil {
-		return apiv1.ImageList{}, err
-	}
-
-	names, err := remote.List(repo, opts...)
-	if tErr, ok := err.(*transport.Error); ok && tErr.StatusCode == http.StatusNotFound {
-		return apiv1.ImageList{}, nil
-	}
-	if err != nil {
-		return apiv1.ImageList{}, err
-	}
-
-	tags, err := tags2.Get(ctx, s.client, ns)
+	err := s.client.List(ctx, imageList, &kclient.ListOptions{
+		Namespace: image.Namespace,
+	})
 	if err != nil {
-		return apiv1.ImageList{}, err
+		result = append(result, field.InternalError(field.NewPath("namespace"), err))
 	}
 
-	result := apiv1.ImageList{}
-	for _, imageName := range names {
-		if !tags2.SHAPattern.MatchString(imageName) {
+	for _, imageItem := range imageList.Items {
+		if imageItem.Digest == image.Digest {
 			continue
 		}
-		tags := tags[imageName]
-		if len(tags) == 0 {
-			tags = append(tags, "")
-		}
-		for _, tag := range tags {
-			image := apiv1.Image{
-				ObjectMeta: metav1.ObjectMeta{
-					Name:      imageName,
-					Namespace: ns,
-				},
-				Digest:    "sha256:" + imageName,
-				Reference: tag,
+		for i, tag := range imageItem.Tags {
+			currentTag := imageItem.Tags[i]
+			if duplicateTag[imageItem.Tags[i]] {
+				result = append(result, field.Duplicate(field.NewPath("tag name"), fmt.Errorf("unable to tag image %s with tag %s as it is already in use by %s", image.Name[:12], tag, imageItem.Name[:12])))
 			}
-			if tag != "" {
-				parsedTag, err := name.NewTag(tag)
-				if err == nil {
-					image.Repository = strings.TrimSuffix(tag, ":"+parsedTag.TagStr())
-					image.Tag = parsedTag.TagStr()
-				}
-			}
-			result.Items = append(result.Items, image)
+			duplicateTag[currentTag] = true
 		}
 	}
+	return result
+}
 
-	return result, nil
+func (s *Strategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) (result field.ErrorList) {
+	return s.Validate(ctx, obj)
 }
 
-func (s *Strategy) Delete(ctx context.Context, obj types.Object) (types.Object, error) {
-	image, matchedReference, err := s.imageGet(ctx, obj.GetNamespace(), obj.GetName())
-	if err != nil {
-		return nil, err
-	}
+func (s *Strategy) ConvertToTable(ctx context.Context, object runtime.Object, tableOptions runtime.Object) (*metav1.Table, error) {
+	return tables.ImageConverter.ConvertToTable(ctx, object, tableOptions)
+}
+func getRepo(namespace string) (name.Repository, error) {
```

Comment:
Mentioned this previously elsewhere, but reiterating because @ibuildthecloud just noticed it too:
we really prefer a new line between functions. Please make this a common practice in all your code.

So,
```
func foo() {
    ....
}

func bar() {
    ...
}
```
not
```
func foo() {
    ....
}
func bar() {
    ...
}
```

Reply:
went through should all be spaced now lmk if any were missed